Source/Core/VideoCommon/PixelEngine.cpp

```diff
@@ -277,22 +277,29 @@ static void SetTokenFinish_OnMainThread(u64 userdata, s64 cyclesLate)
 // Raise the event handler above on the CPU thread.
 // s_token_finish_mutex must be locked.
 // THIS IS EXECUTED FROM VIDEO THREAD
-static void RaiseEvent()
+static void RaiseEvent(int cycles_into_future)
 {
   if (s_event_raised)
     return;
 
   s_event_raised = true;
 
   CoreTiming::FromThread from = CoreTiming::FromThread::NON_CPU;
+  s64 cycles = 0;  // we don't care about timings for dual core mode.
   if (!SConfig::GetInstance().bCPUThread || Fifo::UseDeterministicGPUThread())
+  {
     from = CoreTiming::FromThread::CPU;
-  CoreTiming::ScheduleEvent(0, et_SetTokenFinishOnMainThread, 0, from);
+
+    // Hack: Dolphin's single-core gpu timings are way too fast. Enforce a minimum delay to give
+    //       games time to setup any interrupt state
+    cycles = std::max(500, cycles_into_future);
+  }
+  CoreTiming::ScheduleEvent(cycles, et_SetTokenFinishOnMainThread, 0, from);
```

Comment:
I'm noticing a pre-existing bug here, which I fear I'm making worse.

There is no synchronisation, so if there are many finish or tokens close to each other in the FIFO, they will stomp over each other. The cpu thread will only see a single interrupt, and the token will jump straight to the final value.

This can happen in dual-core mode, randomly if the gpu thread is fast, or the cpu thread is slow.  
Before this PR, this can happen in single core mode, but only within a single FIFO window (or if multiple gather-pipe bursts happen within the same JIT block)
After this PR, this can also happen when multiple gather-pipe bursts happen to within 500 cycles of each other.

I fear that this bug needs be fixed before merging this fix, otherwise we risk causing more regressions.